python/metatensor-learn/metatensor/learn/nn/equivariant_transform.py

```diff
@@ -0,0 +1,145 @@
+    :param invariant_key_idxs: list of int, the indices of the invariant keys present in
+        `in_keys` in the input :py:class:`TensorMap`. Only blocks for these keys will
+        have bias applied according to the user choice. Covariant blocks will not have
+        bias applied.
+    :param in_features: list of int, the number of features in the input tensor for each
+        block indexed by the keys in :param in_keys:. If passed as a single value, the
+        same number of features is assumed for all blocks.
+    :param out_properties: list of :py:class`Labels` (optional), the properties labels
+        of the output. By default the output properties are relabeled using
+        Labels.range.
+    :param device: :py:class:`str` or :py:class:`torch.device`, the computational device
+        to use for calculations.
+    :param dtype: :py:class:`torch.dtype` , the scalar type to use to store parameters.
+    """
+
+    def __init__(
+        self,
+        module: callable,
+        in_keys: Labels,
+        invariant_key_idxs: List[int],
+        in_features: Union[int, List[int]],
+        out_features: Optional[Union[int, List[int]]] = None,
+        out_properties: Optional[List[Labels]] = None,
+        *,
+        device: Optional[torch.device] = None,
+        dtype: Optional[torch.dtype] = None,
+    ) -> None:
+        super().__init__()
+        # Infer `out_features` if not provided
+        if out_features is None:
+            if out_properties is None:
+                raise ValueError(
+                    "If `out_features` is not provided,"
+                    " `out_properties` must be provided."
+                )
+            out_features = [len(p) for p in out_properties]
+
+        # Check input parameters, convert to lists (for each key) if necessary
+        in_features = _check_module_map_parameter(
+            in_features, "in_features", int, len(in_keys), "in_keys"
+        )
+
+        modules: List[Module] = []
+        for i in range(len(in_keys)):
+            if i in invariant_key_idxs:
+                module_i = module
+            else:
+                module_i = _CovariantTransform(
+                    in_features=in_features[i],
+                    module=module,
+                    device=device,
+                    dtype=dtype,
+                )
+            modules.append(module_i)
+
+        self.module_map = ModuleMap(in_keys, modules, out_properties)
+
+    def forward(self, tensor: TensorMap) -> TensorMap:
+        """
+        Apply the transformation to the input tensor map `tensor`.
+
+        :param tensor: :py:class:`TensorMap` with the input tensor to be transformed.
+        :return: :py:class:`TensorMap`
+        """
+        return self.module_map(tensor)
+
+
+class _CovariantTransform(Module):
+    """
+    Applies an arbitrary shape-preserving transformation defined in ``module`` to a
+    3-dimensional tensor in a way that preserves equivariance.
```

Comment:
this needs to say it is using the norm to create invariants

Reply:
@ppegolo just a reminder on this one :)